### PR TITLE
[IOTDB-944] Support UDTF (User-defined Timeseries Generating Function)

### PR DESCRIPTION
**Pull request by SteveYurongSu**
JIRA: https://issues.apache.org/jira/browse/IOTDB-944

The system design doc and user doc are still being written. I will finish the doc as soon as possible and share the links here. :D

### CODE REVIEW

**Comment by SteveYurongSu** [comment]
Product Requirement Doc (Chinese): https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=165221503
Design Doc (Chinese): https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=165220355

File: server/src/main/java/org/apache/iotdb/db/exception/UDFRegistrationException.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.exception;
+
+import org.apache.iotdb.rpc.TSStatusCode;
+
+public class UDFRegistrationException extends IoTDBException {
```
Comment (samperson1997) [inline server/src/main/java/org/apache/iotdb/db/exception/UDFRegistrationException.java:24]: I recommend to make `UDFRegistrationException` extend `StorageEngineException`. In this way, we don't need to change that many files.

**Review by samperson1997 — CHANGES_REQUESTED** [review]

**Comment by SteveYurongSu** [comment]
@HTHou plz invoke the sonar cloud check for this PR :D

**Comment by HTHou** [comment]
> @HTHou plz invoke the sonar cloud check for this PR :D

Done!

**Comment by SteveYurongSu** [comment]
Chinese User doc: https://github.com/SteveYurongSu/iotdb/blob/udf/docs/zh/UserGuide/Operation%20Manual/UDF%20User%20Defined%20Function.md
English User doc: https://github.com/SteveYurongSu/iotdb/blob/udf/docs/UserGuide/Operation%20Manual/UDF%20User%20Defined%20Function.md

File: server/src/main/java/org/apache/iotdb/db/query/udf/service/UDFRegistrationService.java [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+  }
+
+  private URLClassLoader getUDFClassLoader() throws IOException {
+    Collection<File> files = FileUtils
+        .listFiles(SystemFileFactory.INSTANCE.getFile(libRoot), null, true);
+    URL[] urls = FileUtils.toURLs(files.toArray(new File[0]));
+    return new URLClassLoader(urls);
+  }
+
+  public void deregister(String functionName) throws UDFRegistrationException {
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/udf/service/UDFRegistrationService.java:151]: There may be a resource leak here.
I think you should unload the corresponding class in the URLClassLoader when the `deregister` function is called.

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/query/udf/service/UDFRegistrationService.java [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+  private void recoveryFromLogFile(File logFile) throws IOException {
+    HashMap<String, String> recoveredUDFs = new HashMap<>();
+
+    try (BufferedReader reader = new BufferedReader(new FileReader(logFile))) {
+      String line;
+      while ((line = reader.readLine()) != null) {
+        String[] data = line.split(",");
+        short type = Short.parseShort(data[0]);
+        if (type == UDFLogWriter.REGISTER_TYPE) {
+          recoveredUDFs.put(data[1], data[2]);
```
[the remark this answers is not in the document]
Comment (SteveYurongSu) [reply server/src/main/java/org/apache/iotdb/db/query/udf/service/UDFRegistrationService.java:259]: The return value of readLine() method is not including any line-termination characters, so I think my implementation here is ok :D

File: server/src/main/java/org/apache/iotdb/db/query/udf/service/UDFRegistrationService.java [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+        logger.warn(errorMessage);
+        throw new UDFRegistrationException(errorMessage);
+      }
+    }
+
+    Class<?> functionClass;
+    try {
+      udfClassLoader = getUDFClassLoader();
+      functionClass = Class.forName(className, true, udfClassLoader);
+      functionClass.getDeclaredConstructor().newInstance();
```
[the remark this answers is not in the document]
Comment (SteveYurongSu) [reply server/src/main/java/org/apache/iotdb/db/query/udf/service/UDFRegistrationService.java:118]: Yes, it's not necessary. I just want to make sure the instantiation process is fine, should I remove the code?

File: server/src/main/java/org/apache/iotdb/db/query/udf/service/UDFRegistrationService.java [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+            functionName, className,
+            information.getFunctionName(), information.getClassName());
+        logger.warn(errorMessage);
+        throw new UDFRegistrationException(errorMessage);
+      }
+    }
+
+    Class<?> functionClass;
+    try {
+      udfClassLoader = getUDFClassLoader();
```
[the remark this answers is not in the document]
Comment (SteveYurongSu) [reply server/src/main/java/org/apache/iotdb/db/query/udf/service/UDFRegistrationService.java:116]: It seems that the `addURL` method is protected and all URLs should be determined before the construction of class URLClassLoader. I don't know if there's a more elegant way :)

[the remark this answers is not in the document]
Comment (SteveYurongSu) [reply server/src/main/java/org/apache/iotdb/db/query/udf/service/UDFRegistrationService.java:116]: Maybe I can extend the class URLClassLoader to solve the problem. Let me have a try.

**Review by Alima777 — APPROVED** [review]
LGTM

File: server/src/main/java/org/apache/iotdb/db/query/udf/service/UDFRegistrationService.java [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+
+public class UDFRegistrationService implements IService {
+
+  private static final Logger logger = LoggerFactory.getLogger(UDFRegistrationService.class);
+
+  private static final String ULOG_FILE_DIR =
+      IoTDBDescriptor.getInstance().getConfig().getSystemDir()
+          + File.separator + "udf" + File.separator;
+  private static final String LOG_FILE_NAME = ULOG_FILE_DIR + "ulog.txt";
+  private static final String TEMPORARY_LOG_FILE_NAME = LOG_FILE_NAME + ".tmp";
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/query/udf/service/UDFRegistrationService.java:54]: rename?

File: server/src/main/java/org/apache/iotdb/db/query/udf/core/executor/UDTFExecutor.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    collector = ElasticSerializableTVList
+        .newElasticSerializableTVList(configurations.getOutputDataType(), queryId,
+            collectorMemoryBudgetInMB, 1);
+  }
+
+  public void execute(Row row) throws QueryProcessException {
+    try {
+      udtf.transform(row, collector);
+    } catch (Exception e) {
+      throw new QueryProcessException(e.getMessage());
```
Comment (liutaohua) [inline server/src/main/java/org/apache/iotdb/db/query/udf/core/executor/UDTFExecutor.java:64]: Pay attention to the errors, such as when `ArrayIndexOutOfBoundsException` occurs, will only hint `Msg: 500: 1`

File: server/src/main/java/org/apache/iotdb/db/qp/physical/sys/DropFunctionPlan.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+import org.apache.iotdb.db.metadata.PartialPath;
+import org.apache.iotdb.db.qp.logical.Operator.OperatorType;
+import org.apache.iotdb.db.qp.physical.PhysicalPlan;
+
+public class DropFunctionPlan extends PhysicalPlan {
+
+  private final String udfName;
+
+  public DropFunctionPlan(String udfName) {
+    super(false, OperatorType.DROP_FUNCTION);
```
[the remark this answers is not in the document]
Comment (SteveYurongSu) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/sys/DropFunctionPlan.java:33]: The field `canBeSplit ` is useless, maybe it can be removed..

**Comment by SteveYurongSu** [comment]
Hi, I think this PR is ready to be merged. Please take a look again:)

**Review by JackieTien97 — APPROVED** [review]

**Review by samperson1997 — APPROVED** [review]

**Review by qiaojialin — APPROVED** [review]
A big step...

**Pull request merged**